x-pack/platform/plugins/shared/rule_registry/server/routes/get_alert_fields_by_rule_type_ids.ts

```diff
@@ -0,0 +1,154 @@
+          })
+        ),
+      },
+      security: {
+        authz: {
+          requiredPrivileges: ['rac'],
+        },
+      },
+      options: {
+        access: 'internal',
+      },
+    },
+    async (context, request, response) => {
+      try {
+        const { ruleTypeIds = [] } = request.query;
+        const racContext = await context.rac;
+
+        const alertsClient = await racContext.getAlertsClient();
+
+        const authorization = await racContext.alerting.getAlertingAuthorizationWithRequest(
+          request
+        );
+
+        const internalUserEsClient = (await context.core).elasticsearch.client.asInternalUser;
+        const currentUserEsClient = (await context.core).elasticsearch.client.asCurrentUser;
+
+        const registeredRuleTypes = racContext.alerting.listTypes();
+
+        // fetch all rule types if no specific rule type Ids are provided
+        const allRuleTypesIds = ruleTypeIds.length
+          ? Array.isArray(ruleTypeIds)
+            ? ruleTypeIds
+            : [ruleTypeIds]
+          : Array.from(registeredRuleTypes.keys());
+
+        const authorizedRuleTypes = await authorization.getAllAuthorizedRuleTypesFindOperation({
+          authorizationEntity: AlertingAuthorizationEntity.Alert,
+          ruleTypeIds: allRuleTypesIds,
+        });
+
+        const authorizedRuleTypesIds = Array.from(authorizedRuleTypes.keys());
+
+        const siemRuleTypeIds = authorizedRuleTypesIds.filter((ruleTypeId) =>
+          isSiemRuleType(ruleTypeId)
+        );
+        const siemIndices =
+          (siemRuleTypeIds ? await alertsClient.getAuthorizedAlertsIndices(siemRuleTypeIds) : []) ??
+          [];
+
+        const otherRuleTypeIds = authorizedRuleTypesIds.filter(
+          (ruleTypeId) => !isSiemRuleType(ruleTypeId)
+        );
+        const otherIndices =
+          (otherRuleTypeIds
+            ? await alertsClient.getAuthorizedAlertsIndices(otherRuleTypeIds)
+            : []) ?? [];
+
+        const indexFilter = {
+          range: {
+            '@timestamp': {
+              gte: 'now-90d',
+            },
+          },
+        };
+
+        const dataViewsService = await racContext.dataViews.dataViewsServiceFactory(
+          racContext.savedObjectsClient,
+          currentUserEsClient
+        );
+
+        // Fetch fields for SIEM indices using current user client
+        const siemFields = siemIndices.length
+          ? await dataViewsService.getFieldsForWildcard({
+              pattern: siemIndices.join(','),
+              allowNoIndex: true,
+              includeEmptyFields: false,
+              indexFilter,
+            })
+          : [];
+
+        const indexPatternsFetcherAsInternalUser = new IndexPatternsFetcher(internalUserEsClient);
+
+        // Fetch fields for other indices using internal user client
+        const { fields: otherFields } = otherIndices.length
+          ? await indexPatternsFetcherAsInternalUser.getFieldsForWildcard({
```

Comment:
also this, you could call this in parallel with `await indexPatternsFetcherAsInternalUser.getFieldsForWildcard`. I would think this is where we could win the most time

Reply:
[done](https://github.com/elastic/kibana/pull/222032/commits/42536f500591d8882ced74b54ec30f0e07a80dd8#diff-bbccd80f85af0d00f6eafdbc9e444164e6fee357697dc7be42e978fd376a357c)